Blinding/src/TrackDigiExtractor_module.cc

```diff
@@ -0,0 +1,122 @@
+#include "Offline/RecoDataProducts/inc/StrawHit.hh"
+#include "Offline/RecoDataProducts/inc/StrawDigi.hh"
+#include "Offline/RecoDataProducts/inc/TrkStrawHitSeed.hh"
+
+namespace mu2e{
+  class TrackDigiExtractor: public art::EDProducer{
+    public:
+      struct Config{
+        // track
+        fhicl::Atom<art::InputTag> kalseed_tag{
+          fhicl::Name("KalSeedCollection"),
+          fhicl::Comment("art::InputTag of KalSeeds to extract digis from")
+        };
+
+        // tracker
+        fhicl::Atom<art::InputTag> strawhit_tag{
+          fhicl::Name("StrawHitCollection"),
+          fhicl::Comment("art::InputTag of StrawHits associated with KalSeed")
+        };
+        fhicl::Atom<art::InputTag> strawdigi_tag{
+          fhicl::Name("StrawDigiCollection"),
+          fhicl::Comment("art::InputTag of StrawDigis associated with StrawHits")
+        };
+
+        // calorimeter...
+      };
+
+      using Parameters = art::EDProducer::Table<Config>;
+      TrackDigiExtractor(const Parameters&);
+
+    protected:
+      art::InputTag _kalseed_tag;
+      art::InputTag _strawhit_tag;
+      art::InputTag _strawdigi_tag;
+
+    private:
+      void produce(art::Event&);
+  };
+
+  TrackDigiExtractor::TrackDigiExtractor(const Parameters& config):
+      art::EDProducer(config),
+      _kalseed_tag(config().kalseed_tag()),
+      _strawhit_tag(config().strawhit_tag()),
+      _strawdigi_tag(config().strawdigi_tag()){
+    this->consumes<KalSeedCollection>(_kalseed_tag);
+    this->consumes<StrawHitCollection>(_strawhit_tag);
+    this->consumes<StrawDigiCollection>(_strawdigi_tag);
+    this->consumes<StrawDigiADCWaveformCollection>(_strawdigi_tag);
+    this->produces<StrawDigiCollection>();
+    this->produces<StrawDigiADCWaveformCollection>();
+  }
+
+  void TrackDigiExtractor::produce(art::Event& event){
+    // data-product lookups
+    auto kalseeds = event.getValidHandle<KalSeedCollection>(_kalseed_tag);
+    auto strawhits = event.getValidHandle<StrawHitCollection>(_strawhit_tag);
+    auto strawdigis = event.getValidHandle<StrawDigiCollection>(_strawdigi_tag);
+    auto strawadcss = event.getValidHandle<StrawDigiADCWaveformCollection>(_strawdigi_tag);
+
+    // sanity checks
+    if (strawhits->size() != strawdigis->size()){
+      std::string msg = "mismatched StrawHit and StrawDigi collection lengths";
+      throw cet::exception("TrackDigiExtractor") << msg << std::endl;
+    }
+    if (strawdigis->size() != strawadcss->size()){
+      std::string msg = "mismatched StrawDigi and StrawDigiADCWaveform collection lengths";
+      throw cet::exception("TrackDigiExtractor") << msg << std::endl;
+    }
+
+    // deep-copy digis underlying tracks
+    auto ext_strawdigis = std::make_unique<StrawDigiCollection>();
+    auto ext_strawadcss = std::make_unique<StrawDigiADCWaveformCollection>();
+    for (const auto& kalseed: *kalseeds){
+      // tracker
+      auto seeds = kalseed.hits();
+      for (const auto& seed: seeds){
```

Comment:
Its not clear to me we want to keep the inactive hits in this use case.

Reply:
The only place I can identify this being set is in the `TrkStrawHitSeed` constructor, where it propagates from a `WireHitState` --- it looks like this is still a fit-level construct (e.g., it contains the decision on left/right ambiguity), so is contextual to the fit. If that's the case, then I think we'd want to keep the inactive hits, as they could become active when fit in the presence of pileup. Is that a reasonable take? 

Or, if I'm wrong about the above, then we have to ask whether only propagating active hits risks a break in blindness. Since these are overlaid onto pileup which could contribute hits to the fit but end up inactive, the answer may be no, but it's worth mentioning.

Reply:
In OffSpill cosmic track events, the inactive hits will mostly be poorly measured hits created by the primary particle, so you want to keep them. However in a mixed event, many/most of the inactive hits will be from pileup, so you would not want to keep them if you planned to re-mix the extracted digis.
I think the current code covers 90% of the foreseeable use cases, lets go with this for now. If in future we want to use this code to extract digis from events with pileup we can extend it to include a hit selection algorithm.

Reply:
Ok, that all makes sense.